apps/communications/contacts/js/bootstrap.js

```diff
@@ -0,0 +1,267 @@
+      return;
+    }
+
+    localStorage.setItem(aCache.id, JSON.stringify(aCache.content));
+  }
+
+  function deleteCache(aId) {
+    localStorage.removeItem(aId);
+  }
+
+  function getCachedContacts(aNodeList) {
+    if (!aNodeList) {
+      return;
+    }
+
+    if (!_cachedContacts) {
+      _cachedContacts = new Map();
+    }
+
+    for (var i = 0; i < aNodeList.length; i++) {
+      _cachedContacts.set(aNodeList[i].dataset.uuid,
+                          aNodeList[i].innerHTML);
+    }
+  }
+
+  function appendNodesToContainer(aContainer, aNodeList) {
+    if (!aNodeList.length) {
+      return Promise.resolve();
+    }
+
+    return new Promise((resolve) => {
+      var resolved = false;
+      var fragment = document.createDocumentFragment();
+      aNodeList.forEach((node) => {
+        if (!node.elementName) {
+          return;
+        }
+        var child = document.createElement(node.elementName);
+        node.attributes && node.attributes.forEach((attribute) => {
+          if (!attribute.name || !attribute.value) {
+            return;
+          }
+          child.setAttribute(attribute.name, attribute.value);
+        });
+        if (node.innerHTML) {
+          child.innerHTML = node.innerHTML;
+        }
+        fragment.appendChild(child);
+
+        var headerName = child.id.split('section-group-')[1];
+        if (!_cachedHeaders) {
+          _cachedHeaders = {};
+        }
+        _cachedHeaders[headerName] =
+          fragment.querySelector('ol#contacts-list-' + headerName);
+
+        if (!resolved) {
+          resolve();
+        }
+      });
+      getCachedContacts(fragment.querySelectorAll('li[data-cache=true]'));
+      aContainer.appendChild(fragment);
+    });
+  }
+
+  // Exported for testing purposes only.
+  exports.applyCache = function applyCache(aCacheId) {
+    if (!_caches.has(aCacheId)) {
+      return Promise.resolve();
+    }
+
+    var cache = _caches.get(aCacheId);
+
+    var cacheContent = localStorage.getItem(aCacheId);
+    if (!cacheContent) {
+      return Promise.resolve();
+    }
+
+    try {
+      cacheContent = JSON.parse(cacheContent);
+      cache.content = cacheContent;
+      _caches.set(aCacheId, cache);
+    } catch(e) {
+      console.error(e);
+      return Promise.resolve();
+    }
+
+    if (!cache.containerId) {
+      return Promise.resolve();
+    }
+
+    var container = document.getElementById(cache.containerId);
+    if (!container) {
+      console.warning('Could not apply cached content to ' +
+                      cache.containerId);
+      return Promise.resolve();
+    }
+
+    return appendNodesToContainer(container, cacheContent).then(() => {
+      cache.content = null;
+      cache.active = true;
+      _caches.set(aCacheId, cache);
+      return Promise.resolve();
+    });
+  };
+
+  var Cache = {
+    get active() { 
+      return _caches.get(FIRST_CHUNK).active;
+    },
+
+    get enabled () {
+      // XXX: setup this as part of the configuration
```

Comment:
Will be part of this bug? I think it can make the bootstrap a bit more complex, but maybe we can uplift this to 2.2 easier, if it can be disabled.